magicgui/function_gui.py

```diff
@@ -0,0 +1,286 @@
+        providing args/kwargs to the function call.  Only those provided will override
+        the ones from the gui.  A `called` signal will also be emitted with the results.
+
+        Returns
+        -------
+        result : Any
+            whatever the return value of the original function would have been.
+
+        Examples
+        --------
+        gui = FunctionGui(func, show=True)
+
+        # then change parameters in the gui, or by setting:  gui.param.value = something
+
+        gui()  # calls the original function with the current parameters
+        """
+        sig = self.to_signature()
+        bound = sig.bind(*args, **kwargs)
+        bound.apply_defaults()
+
+        value = self._function(*bound.args, **bound.kwargs)
+        if self._result_widget is not None:
+            self._result_widget.value = value
+
+        return_type = self._return_annotation
+        if return_type:
+            for callback in _type2callback(return_type):
+                callback(self, value, return_type)
+        self.called(value=value)
+        return value
+
+    def __repr__(self) -> str:
+        """Return string representation of instance."""
+        fname = f"{self._function.__module__}.{self._function.__name__}"
+        return f"<FunctionGui {fname}{self.to_signature()}>"
+
+    @property
+    def result_name(self) -> str:
+        """Return a name that can be used for the result of this magicfunction."""
+        return self._result_name or (self._function.__name__ + " result")
+
+    @result_name.setter
+    def result_name(self, value: str):
+        """Set the result name of this MagicGui widget."""
+        self._result_name = value
+
+    def Gui(self, show=False):
+        """Create a widget instance [DEPRECATED]."""
+        warnings.warn(
+            "\n\nCreating a widget instance with `my_function.Gui()` is deprecated,\n"
+            "the magicgui decorator now returns a widget instance directly, so you\n"
+            "should simply use the function itself as a magicgui widget, or call\n"
+            "`my_function.show(run=True)` to run the application.\n"
+            "In a future version, the `Gui` attribute will be removed.\n",
+            FutureWarning,
+        )
+        if show:
+            self.show()
+        return self
+
+
+# ==================   magicgui decorator   ===================================
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+@overload
+def magicgui(function: F, **k) -> FunctionGui:  # noqa: D103
+    ...
+
+
+@overload
+def magicgui(function=None, **k) -> Callable[[F], FunctionGui]:  # noqa: D103
+    ...
+
+
+def magicgui(
+    function: Optional[F] = None,
+    *,
+    orientation: str = "horizontal",
+    labels: bool = True,
+    call_button: Union[bool, str] = False,
+    auto_call: bool = False,
+    result_widget: bool = False,
+    app: AppRef = None,
+    **param_options: dict,
+):
```

Comment:
If I'm not mistaken, this is really the core API for magicgui, which we're planning to allow users to access in https://github.com/napari/napari/pull/1856 via the `magic_kargs` argument?